service.cpp

```diff
@@ -0,0 +1,296 @@
+#include "service.h"
+#include <ctime>
+#include <cstdlib>
+#include <chrono>
+#include <string>
+#include <iostream>
+#include <vector>
+#include <map>
+#include <sstream>
+#include <algorithm>
+#include <iterator>
+using grpc::Server;
+using grpc::ServerBuilder;
+using grpc::ServerContext;
+using grpc::ServerReader;
+using grpc::ServerReaderWriter;
+using grpc::ServerWriter;
+using grpc::Status;
+using grpc::StatusCode;
+using chirp::Chirp;
+using chirp::ChirpReply;
+using chirp::ChirpRequest;
+using chirp::FollowReply;
+using chirp::FollowRequest;
+using chirp::MonitorReply;
+using chirp::MonitorRequest;
+using chirp::ReadReply;
+using chirp::ReadRequest;
+using chirp::RegisterReply;
+using chirp::RegisterRequest;
+using chirp::Timestamp;
+using chirp::ServiceLayer;
+
+const std::string ServiceImpl::CHIRP_ID_KEY = "G_chirpid";
+const std::string ServiceImpl::USER_ID_KEY = "G_userid";
+
+Status ServiceImpl::registeruser(ServerContext* context, 
+                    const RegisterRequest* registerRequest, 
+                    RegisterReply* registerReply) {
+  std::string username = registerRequest->username(); // assume username is not empty
+  std::cout << "[LOG] Attempting to register " << username << std::endl;
+  if (storageclient_.has("USER_"+username)) { // username exists
+    std::cout << "[LOG] " << "username exists" << std::endl;
+    Status status(StatusCode::ALREADY_EXISTS, "existed username");
+    return status;
+  } else {
+    while (true) {
+      std::string userid = storageclient_.get("G_userid");
+      if (userid == storageclient_.get("G_userid")) { //TODO lock this part
+        storageclient_.put("G_userid", std::to_string(std::stoi(userid)+1));
+        storageclient_.put("USER_" + username, userid);
+        std::cout << "[LOG] " << "username registered with id " << userid << std::endl;
+        return Status::OK;
+      }
+    }
+  }
+}
+
+Status ServiceImpl::login(ServerContext* context, 
+                          const RegisterRequest* registerRequest, 
+                          RegisterReply* registerReply) {
+  std::string username = registerRequest->username(); // assume username is not empty
+  if (storageclient_.has("USER_"+username)) { // username exists
+    std::cout << "[LOG] " << username << " signed in" << std::endl;
+    return Status::OK;
+  } else {
+    Status status(StatusCode::NOT_FOUND, "user doesn't exist");
+    return status;
+  }
+}
+
+Status ServiceImpl::chirp(ServerContext* context, const ChirpRequest* chirpRequest, ChirpReply* chirpReply) {
```

Comment:
Since this function also add two additional fields, consider making this function a transaction? (If failed in any one of the operation, reset everything to avoid the scenario when you have insert the chirp but it will not appear in `userchirps`)